pandas/core/reshape.py

```diff
@@ -940,24 +942,27 @@ def melt_stub(df, stub, i, j):
         varnames = get_var_names(df, "^" + stub)
         newdf = melt(df, id_vars=i, value_vars=varnames, value_name=stub,
                      var_name=j)
-        newdf_j = newdf[j].str.replace(stub, "")
-        try:
-            newdf_j = newdf_j.astype(int)
-        except ValueError:
-            pass
-        newdf[j] = newdf_j
-        return newdf
+        newdf[j] = Categorical(newdf[j])
+        newdf[j] = newdf[j].str.replace(stub, "")
+        return newdf.set_index([i, j])
 
     id_vars = get_var_names(df, "^(?!%s)" % "|".join(stubnames))
     if i not in id_vars:
         id_vars += [i]
 
-    newdf = melt_stub(df, stubnames[0], id_vars, j)
+    mstubs = [melt_stub(df, stub, i, j) for stub in stubnames]
+    mstubs = mstubs[0].join(mstubs[1:], how='outer')
+    new = df[id_vars].set_index(i).join(mstubs)
+
+    # The index of the new dataframe is [i, j], if the j column is a time
+    # variable, try to convert this to integer.
```

Comment:
I don't like the auto coercing of the strings -> ints. This is not very idiomatic and unexpected. I would leave the columns as strings.